src/docfx/lib/JsonUtility.cs

```diff
@@ -125,6 +121,103 @@ public static JObject Merge(JObject first, IEnumerable<JObject> objs)
             return result;
         }
 
+        public static (List<Error>, JToken) ValidateNullValue(this JToken token, JTokenSourceMap mappings = null)
+        {
+            var errors = new List<Error>();
+            var nullNodes = new List<JToken>();
+            token.Traverse(errors, mappings, nullNodes);
+            foreach (var node in nullNodes)
+            {
+                node.Remove();
+            }
+            return (errors, token);
+        }
+
+        private static bool IsNullOrUndefined(this JToken token)
+        {
+            return
+                (token == null) ||
+                (token.Type == JTokenType.Null) ||
+                (token.Type == JTokenType.Undefined);
+        }
+
+        /// <summary>
+        /// Parse a string to JToken.
+        /// Validate null value during the process.
+        /// </summary>
+        private static (List<Error>, JToken) Deserialize(TextReader reader)
+        {
+            using (JsonReader json = new JsonTextReader(reader))
+            {
+                return DefaultDeserializer.Deserialize<JToken>(json).ValidateNullValue();
+            }
+        }
+
+        private static void Traverse(this JToken token, List<Error> errors, JTokenSourceMap mappings, List<JToken> nullNodes)
```

Comment:
this can be a single function, because we only care about 3 `JToken` types: `JObject`, `JArray` and `JValue`, the skeleton can be:

```csharp
void traverse(JToken token)
{
    if (token is JArray arr) {}
    else if (token is JObject obj) {}
    else if (token is JValue v) {}
}
```

Reply:
Right, but we only care about JArray and JObject, since if the value of JObject is null, we will log warning and remove this JObject.

---
In reply to: [200283844](https://github.com/dotnet/docfx/pull/2994#discussion_r200283844) [](ancestors = 200283844)